### PR TITLE
Whales - Sarah Williams

### PR DESCRIPTION
**Pull request by sarah-kw**
viewing-party project submission

### CODE REVIEW

File: tests/test_wave_01.py [hunk 1]
```diff
@@ -116,14 +116,15 @@ def test_moves_movie_from_watchlist_to_empty_watched():
     updated_data = watch_movie(janes_data, MOVIE_TITLE_1)
 
     # Assert
-    assert len(updated_data["watchlist"]) is 0
-    assert len(updated_data["watched"]) is 1
-    
+    assert len(updated_data["watchlist"]) == 0
+    assert len(updated_data["watched"]) == 1
+    assert ({"title": MOVIE_TITLE_1, "genre": GENRE_1, "rating": RATING_1}
+        in updated_data["watched"])
```
Comment (alope107) [inline tests/test_wave_01.py:122]: Nice asserts!

File: tests/test_wave_01.py [hunk 2]
```diff
@@ -139,14 +140,15 @@ def test_moves_movie_from_watchlist_to_watched():
     updated_data = watch_movie(janes_data, movie_to_watch["title"])
 
     # Assert
-    assert len(updated_data["watchlist"]) is 1
-    assert len(updated_data["watched"]) is 2
+    assert len(updated_data["watchlist"]) == 1
+    assert len(updated_data["watched"]) == 2
+    assert HORROR_1 in updated_data["watched"]
```
Comment (alope107) [inline tests/test_wave_01.py:145]: These asserts work great! One way you could make this slightly more robust is by checking that the entire list matches what you expected. This makes sure to catch any bugs where the first element in the list was accidentally overwritten or modified.

File: tests/test_wave_02.py [hunk 3]
```diff
@@ -50,4 +50,22 @@ def test_genre_is_None_if_empty_watched():
...
+#     # Arrange
+#     janes_data = {
+#         "watched": []
+#     }
+
+#     # Act
+#     popular_genre = get_most_watched_genre(janes_data)
+
+#     # Assert
+#     assert popular_genre == ## List? longer string?
```
Comment (alope107) [inline tests/test_wave_02.py:71]: Love that you're thinking of extra edge cases you could test!

File: viewing_party/party.py [hunk 4]
```diff
@@ -1,23 +1,134 @@
 # ------------- WAVE 1 --------------------
 
 def create_movie(title, genre, rating):
-    pass
+    if not title or not genre or not rating:
+        return None
```
Comment (alope107) [inline viewing_party/party.py:5]: Great input validation!

File: viewing_party/party.py [hunk 5]
```diff
@@ -1,23 +1,134 @@
...
+    return user_data
+
+def watch_movie(user_data, movie_title):
+    user_watchlist = user_data["watchlist"]
+    watching_index = None
+
+    for index, movie in enumerate(user_watchlist):
+        if movie["title"] == movie_title:
+            watching_index = index
+            break
```
Comment (alope107) [inline viewing_party/party.py:24]: Great use of `enumerate`!

File: viewing_party/party.py [hunk 6]
```diff
@@ -1,23 +1,134 @@
...
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
+def get_watched_avg_rating(user_data):
+    watched_list = user_data["watched"]
+    rating_sum = 0
+    number_of_ratings = len(watched_list)
+    if number_of_ratings == 0:
+        return 0
```
Comment (alope107) [inline viewing_party/party.py:39]: Good job preventing a divide by zero error.

File: viewing_party/party.py [hunk 7]
```diff
@@ -1,23 +1,134 @@
...
+    watched_list = user_data["watched"]
+    if not watched_list:
+        return None
+    genre_counter = {}
+    for movie in watched_list:
+        genre = movie["genre"]
+        if genre in genre_counter.keys():
+            genre_counter[genre] += 1
+        else:
+            genre_counter[genre] = 1
```
Comment (alope107) [inline viewing_party/party.py:56]: Nice use of a frequency dictionary!

File: viewing_party/party.py [hunk 8]
```diff
@@ -1,23 +1,134 @@
...
+    return rating_sum/number_of_ratings
+
+def get_most_watched_genre(user_data):
+    watched_list = user_data["watched"]
+    if not watched_list:
+        return None
+    genre_counter = {}
+    for movie in watched_list:
+        genre = movie["genre"]
+        if genre in genre_counter.keys():
```
Comment (alope107) [inline viewing_party/party.py:53]: One small nitpick: the `.keys()` is unnecessary here. For a dict, `in` checks the keys by default. You can simply write:
```
if genre in genre_counter:
```

File: viewing_party/party.py [hunk 9]
```diff
@@ -1,23 +1,134 @@
...
+            genre_counter[genre] += 1
+        else:
+            genre_counter[genre] = 1
+    
+    most_watched_genre = max(genre_counter, key=genre_counter.get)
```
Comment (alope107) [inline viewing_party/party.py:58]: Nice use of `max`! As extra practice, think through how you could do this using a normal for-loop without `max`.

File: viewing_party/party.py [hunk 10]
```diff
@@ -1,23 +1,134 @@
...
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
-# -----------------------------------------
+# ----------------------------------------- 
+def get_unique_watched(user_data):
+    friends = user_data["friends"]
+    user_movie_list = user_data["watched"]
+    unique_movie_names = set()
+    for movie in user_movie_list:
+        unique_movie_names.add(movie["title"])
```
Comment (alope107) [inline viewing_party/party.py:69]: Very clever to only put the title in the `set`! It wasn't an issue in this project, but can you think of what might happen here if there were multiple different movies with the same title? (e.g. there are multiple movies all named "Hamlet")

File: viewing_party/party.py [hunk 11]
```diff
@@ -1,23 +1,134 @@
...
+    
+    for friend in friends:
+        friend_watched_list = friend["watched"]
+        for movie in friend_watched_list:
+            if movie["title"] in unique_movie_names:
+                unique_movie_names.remove(movie["title"])
+
+    unique_movies = [movie for movie in user_movie_list if movie["title"] in unique_movie_names]
```
Comment (alope107) [inline viewing_party/party.py:77]: Nice comprehension!

File: viewing_party/party.py [hunk 12]
```diff
@@ -1,23 +1,134 @@
...
-        
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
+def get_available_recs(user_data):
+    possible_recs = get_friends_unique_watched(user_data)
+
+    available_recs = [rec for rec in possible_recs if rec["host"] in user_data["subscriptions"]]
+    
+    return available_recs
```
Comment (alope107) [inline viewing_party/party.py:108]: Very cool seeing you re-use your code as a helper function!

File: viewing_party/party.py [hunk 13]
```diff
@@ -1,23 +1,134 @@
...
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
 
+def get_new_rec_by_genre(user_data):
+    possible_recs = get_friends_unique_watched(user_data)
+    most_frequent_genre = get_most_watched_genre(user_data)
+
+    genre_recs = [rec for rec in possible_recs if rec["genre"] == most_frequent_genre]
+    
+    return genre_recs
```
Comment (alope107) [inline viewing_party/party.py:121]: Great compact function here!

**Review by alope107 — COMMENTED** [review]
Great job Sarah! This project is a verdant green. Love seeing how frequently you're committing and the clever way you used sets. Very nicely done!
